Librerie/libpdel-0.6.1/pdel/structs/structs.3
.\" @COPYRIGHT@
.\"
.\" Author: Archie Cobbs <archie@freebsd.org>
.\"
.\" $Id: structs.3 901 2004-06-02 17:24:39Z archie $
.\"
.Dd April 22, 2002
.Dt STRUCTS 3
.Os
.Sh NAME
.Nm structs
.Nd library for data structure introspection
.Sh LIBRARY
PDEL Library (libpdel, \-lpdel)
.Sh SYNOPSIS
.In sys/types.h
.In pdel/structs/structs.h
.Ft int
.Fn structs_init "const struct structs_type *type" "const char *name" "void *data"
.Ft int
.Fn structs_reset "const struct structs_type *type" "const char *name" "void *data"
.Ft int
.Fn structs_free "const struct structs_type *type" "const char *name" "void *data"
.Ft int
.Fn structs_equal "const struct structs_type *type" "const char *name" "const void *data1" "const void *data2"
.Ft const struct structs_type *
.Fn structs_find " const struct structs_type *type" "const char *name" "void **datap" "int set_union"
.Ft int
.Fn structs_get "const struct structs_type *type" "const char *name" "const void *from" "void *to"
.Ft int
.Fn structs_set "const struct structs_type *type" "const void *from" "const char *name" "void *to"
.Ft char *
.Fn structs_get_string "const struct structs_type *type" "const char *name" "const void *data" "const char *mtype"
.Ft int
.Fn structs_set_string "const struct structs_type *type" "const char *name" "const char *ascii" "void *data" "char *ebuf" "size_t emax"
.Ft int
.Fn structs_get_binary "const struct structs_type *type" "const char *name" "const void *data" "const char *mtype" "struct structs_data *code"
.Ft int
.Fn structs_set_binary "const struct structs_type *type" "const char *name" "const struct structs_data *code" "void *data" "char *ebuf" "size_t emax"
.Ft int
.Fn structs_traverse "const struct structs_type *type" "const void *data" "char ***listp" "const char *mtype"
.Sh DESCRIPTION
.\"
.Ss Overview
.\"
The
.Nm structs
library includes macros and functions for defining and using
.Nm structs
types.
A
.Nm structs
type is a C structure that contains information describing
some other C data structure.
This information can be used to access the contents of
the described data structure dynamically at run time.
The library provides several pre-defined types for commonly used
data structures, as well as macros for creating new types.
.Pp
A data structure is supported by the
.Nm structs
library if it can be described by a
.Nm structs
type (see
.Xr structs_type 3) .
There are two classes of types: primitive and complex.
Primitive types describe things such as integers, strings, etc.
They are user-definable, and several predefined primitive types are
supplied with the
.Nm structs
library.
Any data structure can be described by a primitive
.Nm structs
type if it has the following properties:
.Pp
.Bl -bullet -offset 3n -compact
.It
It has a fixed size known at compile time.
.It
It can be initialized, uninitialized, copied, and compared for equality.
.It
It can be converted into an ASCII string and back without losing information.
.It
It can be converted into a byte-order independent, self-delimiting
binary sequence and back without losing information.
.El
.Pp
The complex types are defined recursively in terms of other types,
and include the following:
.Pp
.Bl -enum -offset 3n -compact
.It
Pointers
.It
Fixed length arrays
.It
Variable length arrays
.It
Structures
.It
Unions
.Pp
.El
The complex types support accessing sub-elements dircectly by
name at run-time.
That is, array, structure, and union elements can be
accessed by field name or array index expressed as an ASCII string.
The accessed elements may be arbitrarily deep in the data structure.
.Pp
The upshot of all this is that if one takes the time to describe a data
structures with a
.Nm structs
type, then the following operations can be performed dynamically
and automatically on any instance of that data structure:
.Pp
.Bl -bullet -offset 3n -compact
.It
Initialization and uninitialization, including allocating
and freeing heap memory or other resources.
.It
Comparison of two instances for equality
.It
.Dq Deep
copying, i.e., creating a completely new instance
that is a copy of an original with no shared components.
.It
Access to arbitrary sub-fields by name (aka.
.Dq introspection
).
.It
Conversion to/from ASCII (primitive types only)
.It
Conversion to/from XML, with precise input validation
.It
Conversion to/from XML-RPC "values"
.It
Conversion to/from a byte-order independent, self-delimiting byte sequence
.El
.Pp
.\"
.Ss Data Structure Initialization
.\"
A "data structure" is just a contiguous block of memory.
It may of course contain other sub-structures within it,
including pointers to yet other data structures, but for the purposes of the
.Nm structs
library a "data structure" just a block of memory that you can point to.
.Pp
Such a data structure can be in one of two states: uninitialized
or initialized.
For example, a region of heap memory freshly returned by
.Xr malloc 3
is unintialized.
The only valid
.Nm structs
operation on an uninitialized data structure is to initialize it;
this is done by invoking
.Fn structs_init
(see below).
.Pp
Initializing a data structure puts it in a known, valid, default state.
This may involve more than just filling the region of memory with zeros.
For example, it may cause additional heap memory to be allocated
(and initialized), hidden reference counts to be incremented, or other
resources to be allocated.
.Pp
Note that
.Fn structs_init
does not itself allocate the block of memory in which the data structure
is stored, it only initializes it.
The user code must handle allocation of the block of memory.
As a consequence, this memory may live on the stack, or the heap.
Any data structures that are stored in stack variables and are initialized
during execution of a function must be uninitialized before the function
returns to avoid resource leaks.
.Pp
.Fn structs_free
(see below)
is used to free any resources associated with an initialized data structure
and return it to the uninitialized state.
Note that this does not invoke
.Xr free 3
on the block of memory containing the data structure, though it may cause
.Xr free 3
to be invoked for any additional memory previously allocated by
.Fn structs_init .
.\"
.Ss Structs Functions
.\"
Generally speaking, in the functions shown above
.Fa type
points to the
.Nm structs
type describing a data structure,
.Fa data
points to an instance of that data structure, and
.Fa name
references by name the target sub-field or sub-element of the data structure
on which the operation is to take place.
If
.Fa name
is equal to
.Dv NULL
or the empty string then the entire data structure is the target.
In practice,
.Fa name
is often
.Dv NULL .
.Pp
.Fn structs_init
initializes the uninitialized sub-field
.Fa name
of the data structure pointed to by
.Fa data .
The data structure will be set to its default value, which is defined by
.Fa type.
.Pp
.Fn structs_reset
resets the already initialized sub-field
.Fa name
of the data structure pointed to by
.Fa data
to its default value, i.e., the same value that it would have after
a call to
.Fn structs_init .
.Pp
.Fn structs_free
uninitializes the sub-field
.Fa name
of the data structure pointed to by
.Fa data ,
freeing any resources previously allocated by
.Fn structs_init .
.Pp
.Fn structs_equal
compares the sub-fields
.Fa name
of the two data structures pointed to by
.Fa data1
and
.Fa data2
for equality.
It returns 1 if they are equal or 0 if not.
.Pp
.Fn structs_find
locates a sub-field of a data structure by name and returns its
.Nm structs
type.
When invoked,
.Fa "*datap"
should point to the data structure being searched.
Upon successful return, it will point to the sub-field named by
.Fa name .
If
.Fa set_union
is non-zero, then if during the search any unions are encountered
and the union's current field is different from the named field,
then the union's field is changed to the named field and its value
reset to the default value before continuing with the search.
.Pp
.Fn structs_get
generates a copy of the sub-field
.Fa name
in the data structure pointed to by
.Fa from
and places it in the uninitialized region of memory pointed to by
.Fa to ;
.Fa type
is the
.Nm structs
type of
.Fa from .
This is a recursive, or "deep" copy containing no shared elements with
.Fa from .
Note that the
.Nm structs
type of
.Fa "from.<name>"
and
.Fa to
must be the same.
Upon successful return,
.Fa to
will be initialized and therefore it is the caller's responsibility
to eventually uninitialize it.
.Pp
.Fn structs_set
changes the contents of the already initialized sub-field
.Fa name
in the data structure pointed to by
.Fa to
to be a copy of the data structure pointed to by
.Fa from ;
.Fa type
is the
.Nm structs
type of
.Fa to .
This is a recursive, or "deep" copy containing no shared elements with
.Fa from .
Note that the
.Nm structs
type of
.Fa "from"
and
.Fa "to.<name>"
must be the same.
.Fn structs_set
does not modify
.Fa from
in any way.
.Pp
.Fn structs_get_string
returns the ASCII form of the sub-field
.Fa name
in the data structure pointed to by
.Fa data .
This operation is only required to be implemented for primitive types.
The returned string is allocated with
.Xr typed_mem 3
type
.Fa mtype ,
and the caller is responsible for eventually freeing it.
.Pp
.Fn structs_set_string
changes the contents of the already initialized sub-field
.Fa name
in the data structure pointed to by
.Fa data
to the value represented by the ASCII string
.Fa ascii .
This operation is only required to be implemented for primitive types.
If there is an error, e.g.,
.Fa ascii
is not a valid representation of the type, then
.Fn structs_set_string
will return -1 and if
.Fa ebuf
is not
.Dv NULL
an error message (including terminating '\\0') will be printed into the buffer
.Fa ebuf ,
which is assumed to have length
.Fa emax .
.Pp
.Fn structs_get_binary
and
.Fn structs_set_binary
are similar, except that they work with byte-order independent,
self-delimiting binary data instead of ASCII strings.
.Pp
.Fn structs_get_binary
returns the binary encoding of the sub-field
.Fa name
in the data structure pointed to by
.Fa data .
The
.Fa code
argument is a pointer to a
.Li "struct structs_data" :
.Pp
.Bd -literal -compact -offset 3n
struct structs_data {
    u_int     length;       /* number of bytes */
    u_char    *data;        /* pointer to the bytes */
};
.Ed
.Pp
Upon successful return,
.Fa "code->data"
points to the binary encoding, which has length
.Fa "code->length"
and is allocated with
.Xr typed_mem 3
type
.Fa mtype .
The caller is eventually responsible for freeing
.Fa "code->data" .
.Pp
.Fn structs_set_binary
changes the contents of the already initialized sub-field
.Fa name
in the data structure pointed to by
.Fa data
to the value represented by the byte-order independent, self-delimiting
binary encoding described by
.Fa code .
On success, the actual number of bytes consumed is returned; this will
be less than or equal to
.Fa "code->length" .
If there is an error, e.g.,
the encoding was invalid, then
.Fn structs_set_binary
will return -1 and if
.Fa ebuf
is not
.Dv NULL
an error message (including terminating '\\0') will be printed into the buffer
.Fa ebuf ,
which is assumed to have length
.Fa emax .
.Pp
.Fn structs_traverse
generates a list of the names of all of the "leaf" sub-structures in
the data structure pointed to by
.Fa data ;
these will all have primitive
.Nm structs
type.
It returns the number of elements in the array.
A pointer to the array is stored in the location referenced by
.Fa listp.
Each name in the array, as well as the array itself, is allocated with
.Xr typed_mem 3
type
.Fa mtype .
The caller is responsible for freeing all array elements as well as
the array itself.
.Sh RETURN VALUES
All of the above functions indicate an error condition by returning
either -1 or
.Dv NULL
and setting
.Va errno
to an appropriate value.
.Pp
Whenever there is an error, no partial work is done: the state of
the parameters has not changed, and nothing has been allocated or freed.
.Sh SEE ALSO
.Xr libpdel 3 ,
.Xr structs_type 3 ,
.Xr structs_type_array 3 ,
.Xr structs_type_boolean 3 ,
.Xr structs_type_bpf 3 ,
.Xr structs_type_data 3 ,
.Xr structs_type_dnsname 3 ,
.Xr structs_type_ether 3 ,
.Xr structs_type_float 3 ,
.Xr structs_type_id 3 ,
.Xr structs_type_int 3 ,
.Xr structs_type_ip4 3 ,
.Xr structs_type_null 3 ,
.Xr structs_type_pointer 3 ,
.Xr structs_type_regex 3 ,
.Xr structs_type_string 3 ,
.Xr structs_type_struct 3 ,
.Xr structs_type_time 3 ,
.Xr structs_type_union 3 ,
.Xr structs_xml_input 3 ,
.Xr structs_xmlrpc 3 ,
.Xr typed_mem 3
.Sh HISTORY
The PDEL library was developed at Packet Design, LLC.
.Dv "http://www.packetdesign.com/"
.Sh AUTHORS
.An Archie Cobbs Aq archie@freebsd.org


Librerie/libpdel-0.6.1/pdel/structs/structs.c

/*
 * @COPYRIGHT@
 *
 * Author: Archie Cobbs <archie@freebsd.org>
 */

#include <sys/types.h>

#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <assert.h>
#include <string.h>
#include <ctype.h>
#include <errno.h>

#ifndef PD_BASE_INCLUDED
#include "pdel/pd_base.h"	/* picks up pd_port.h */
#endif
#include "structs/structs.h"
#include "structs/type/int.h"
#include "structs/type/array.h"
#include "structs/type/string.h"
#include "structs/type/struct.h"
#include "structs/type/union.h"
#include "util/typed_mem.h"

/* Special handling for array length as a read-only field */
static const struct structs_type structs_type_array_length = {
	sizeof(u_int),
	"uint",
	STRUCTS_TYPE_PRIMITIVE,
	structs_region_init,
	structs_region_copy,
	structs_region_equal,
	structs_int_ascify,
	structs_notsupp_binify,
	structs_region_encode_netorder,
	structs_notsupp_decode,
	structs_nothing_free,
	{ { (void *)2 }, { (void *)0 } }    /* args for structs_int_ascify */
};

/* Special handling for union field name as a read-only field */
static const struct structs_type structs_type_union_field_name = {
	sizeof(char *),
	"string",
	STRUCTS_TYPE_PRIMITIVE,
	structs_notsupp_init,
	structs_notsupp_copy,
	structs_string_equal,
	structs_string_ascify,
	structs_notsupp_binify,
	structs_notsupp_encode,
	structs_notsupp_decode,
	structs_nothing_free,
	{ { (void *)"union field_name" },    /* args for structs_type_string */
	{ (void *)0 } }
};

/*
 * Initialize an item.
 */
int
structs_init(const struct structs_type *type, const char *name, void *data)
{
	/* Find item */
	if ((type = structs_find(type, name, (const void **) &data, 0)) == NULL)
		return (-1);

	/* Initialize it */
	return ((*type->init)(type, data));
}

/*
 * Reset an item.
 */
int
structs_reset(const struct structs_type *type, const char *name, void *data)
{
	void *temp;

	/* Find item */
	if ((type = structs_find(type, name, (const void **) &data, 0)) == NULL)
		return (-1);

	/* Make a temporary new copy */
	if ((temp = MALLOC(TYPED_MEM_TEMP, type->size)) == NULL)
		return (-1);
	memset(temp, 0, type->size);
	if ((*type->init)(type, temp) == -1) {
		FREE(TYPED_MEM_TEMP, temp);
		return (-1);
	}

	/* Replace existing item, freeing it first */
	(*type->uninit)(type, data);
	memcpy(data, temp, type->size);
	FREE(TYPED_MEM_TEMP, temp);
	return (0);
}

/*
 * Free an item, returning it to its initialized state.
 *
 * If "name" is NULL or empty string, the entire structure is free'd.
 */
int
structs_free(const struct structs_type *type, const char *name, void *data)
{
	const int errno_save = errno;

	/* Find item */
	if ((type = structs_find(type, name, (const void **) &data, 0)) == NULL)
		return (-1);

	/* Free it */
	(*type->uninit)(type, data);
	errno = errno_save;
	return (0);
}

/*
 * Get a copy of an item.
 *
 * If "name" is NULL or empty string, the entire structure is copied.
 *
 * It is assumed that "to" points to a region big enough to hold
 * the copy of the item. "to" will not be free'd first, so it should
 * not already be initialized.
 *
 * Note: "name" is only applied to "from". The structs types of
 * "from.<name>" and "to" must be the same.
 */
int
structs_get(const struct structs_type *type,
	const char *name, const void *from, void *to)
{
	/* Find item */
	if ((type = structs_find(type, name, (const void **) &from, 0)) == NULL)
		return (-1);

	/* Copy item */
	return ((*type->copy)(type, from, to));
}

/*
 * Set an item in a structure.
 *
 * If "name" is NULL or empty string, the entire structure is copied.
 *
 * It is assumed that "to" is already initialized.
 *
 * Note: "name" is only applied to "to". The structs types of
 * "from" and "to.<name>" must be the same.
 */
int
structs_set(const struct structs_type *type,
	const void *from, const char *name, void *to)
{
	void *copy;

	/* Find item */
	if ((type = structs_find(type, name, (const void **) &to, 0)) == NULL)
		return (-1);

	/* Make a new copy of 'from' */
	if ((copy = MALLOC(TYPED_MEM_TEMP, type->size)) == NULL)
		return (-1);
	if ((*type->copy)(type, from, copy) == -1) {
		FREE(TYPED_MEM_TEMP, copy);
		return (-1);
	}

	/* Free overwritten item in 'to' */
	(*type->uninit)(type, to);

	/* Move new item in its place */
	memcpy(to, copy, type->size);

	/* Done */
	FREE(TYPED_MEM_TEMP, copy);
	return (0);
}

/*
 * Get the ASCII form of an item.
 */
char *
structs_get_string(const struct structs_type *type,
	const char *name, const void *data, const char *mtype)
{
	/* Find item */
	if ((type = structs_find(type, name, (const void **) &data, 0)) == NULL)
		return (NULL);

	/* Ascify it */
	return ((*type->ascify)(type, mtype, data));
}

/*
 * Set an item's value from a string.
 *
 * The referred to item must be of a type that supports ASCII encoding,
 * and is assumed to be already initialized.
 */
int
structs_set_string(const struct structs_type *type, const char *name,
	const char *ascii, void *data, char *ebuf, size_t emax)
{
	void *temp;
	char dummy[1];

	/* Sanity check */
	if (ascii == NULL)
		ascii = "";
	if (ebuf == NULL) {
		ebuf = dummy;
		emax = sizeof(dummy);
	}

	/* Initialize error buffer */
	if (emax > 0)
		*ebuf = '\0';

	/* Find item */
	if ((type = structs_find(type, name, (const void **) &data, 1)) == NULL) {
		strlcpy(ebuf, strerror(errno), emax);
		return (-1);
	}

	/* Binify item into temporary storage */
	if ((temp = MALLOC(TYPED_MEM_TEMP, type->size)) == NULL)
		return (-1);
	memset(temp, 0, type->size);
	if ((*type->binify)(type, ascii, temp, ebuf, emax) == -1) {
		FREE(TYPED_MEM_TEMP, temp);
		if (emax > 0 && *ebuf == '\0')
			strlcpy(ebuf, strerror(errno), emax);
		return (-1);
	}

	/* Replace existing item, freeing it first */
	(*type->uninit)(type, data);
	memcpy(data, temp, type->size);
	FREE(TYPED_MEM_TEMP, temp);
	return (0);
}

/*
 * Get the binary encoded form of an item, allocated with type 'mtype'.
 */
int
structs_get_binary(const struct structs_type *type, const char *name,
	const void *data, const char *mtype, struct structs_data *code)
{
	/* Find item */
	if ((type = structs_find(type, name, (const void **) &data, 0)) == NULL) {
		memset(code, 0, sizeof(*code));
		return (-1);
	}

	/* Encode it */
	if ((*type->encode)(type, mtype, code, data) == -1) {
		memset(code, 0, sizeof(*code));
		return (-1);
	}

	/* Done */
	return (0);
}

/*
 * Set an item's value from its binary encoded value.
 */
int
structs_set_binary(const struct structs_type *type, const char *name,
	const struct structs_data *code, void *data, char *ebuf, size_t emax)
{
	void *temp;
	int clen;

	/* Initialize error buffer */
	if (emax > 0)
		*ebuf = '\0';

	/* Find item */
	if ((type = structs_find(type, name, (const void **) &data, 0)) == NULL) {
		strlcpy(ebuf, strerror(errno), emax);
		return (-1);
	}

	/* Decode item into temporary storage */
	if ((temp = MALLOC(TYPED_MEM_TEMP, type->size)) == NULL)
		return (-1);
	memset(temp, 0, type->size);
	if ((clen = (*type->decode)(type, code->data,
	    code->length, temp, ebuf, emax)) == -1) {
		FREE(TYPED_MEM_TEMP, temp);
		if (emax > 0 && *ebuf == '\0')
			strlcpy(ebuf, strerror(errno), emax);
		return (-1);
	}
	assert(clen <= code->length);

	/* Replace existing item, freeing it first */
	(*type->uninit)(type, data);
	memcpy(data, temp, type->size);
	FREE(TYPED_MEM_TEMP, temp);

	/* Done */
	return (clen);
}

/*
 * Test for equality.
 */
int
structs_equal(const struct structs_type *type,
	const char *name, const void *data1, const void *data2)
{
	/* Find items */
	if (structs_find(type, name, (const void **) &data1, 0) == NULL)
		return (-1);
	if ((type = structs_find(type, name, (const void **)  &data2, 0)) == NULL)
		return (-1);

	/* Compare them */
	return ((*type->equal)(type, data1, data2));
}

/*
 * Find an item in a structure.
 */
const struct structs_type *
structs_find(const struct structs_type *type, const char *name,
	const void **datap, int set_union)
{
	const void *data = *datap;
	const char *next;

	/* Empty string means stop recursing */
	if (name == NULL || *name == '\0')
		return (type);

	/* Primitive types don't have sub-elements */
	if (type->tclass == STRUCTS_TYPE_PRIMITIVE) {
		errno = ENOENT;
		return (NULL);
	}

	/* Dereference through pointer(s) */
	while (type->tclass == STRUCTS_TYPE_POINTER) {
		type = type->args[0].v;
		data = *((void **)data);
	}

	/* Get next name component */
	if ((next = strchr(name, STRUCTS_SEPARATOR)) != NULL)
		next++;

	/* Find element of aggregate structure update type and data */
	switch (type->tclass) {
	case STRUCTS_TYPE_ARRAY:
	    {
		const struct structs_type *const etype = type->args[0].v;
		const struct structs_array *const ary = data;
		u_long index;
		char *eptr;

		/* Special handling for "length" */
		if (strcmp(name, "length") == 0) {
			type = &structs_type_array_length;
			data = (void *)&ary->length;
			break;
		}

		/* Decode an index */
		index = strtoul(name, &eptr, 10);
		if (!isdigit(*name)
		    || eptr == name
		    || (*eptr != '\0' && *eptr != STRUCTS_SEPARATOR)) {
			errno = ENOENT;
			return (NULL);
		}
		if (index >= ary->length) {
			errno = EDOM;
			return (NULL);
		}
		type = etype;
		data = (char *)ary->elems + (index * etype->size);
		break;
	    }
	case STRUCTS_TYPE_FIXEDARRAY:
	    {
		const struct structs_type *const etype = type->args[0].v;
		const u_int length = type->args[2].i;
		u_long index;
		char *eptr;

		/* Special handling for "length" */
		if (strcmp(name, "length") == 0) {
			type = &structs_type_array_length;
			data = (void *)&type->args[2].i;
			break;
		}

		/* Decode an index */
		index = strtoul(name, &eptr, 10);
		if (!isdigit(*name)
		    || eptr == name
		    || (*eptr != '\0' && *eptr != STRUCTS_SEPARATOR)) {
			errno = ENOENT;
			return (NULL);
		}
		if (index >= length) {
			errno = EDOM;
			return (NULL);
		}
		type = etype;
		data = (char *)data + (index * etype->size);
		break;
	    }
	case STRUCTS_TYPE_STRUCTURE:
	    {
		const struct structs_field *field;

		/* Find the field */
		for (field = type->args[0].v; field->name != NULL; field++) {
			const size_t fnlen = strlen(field->name);

			/* Handle field names with separator in them */
			if (strncmp(name, field->name, fnlen) == 0
			    && (name[fnlen] == '\0'
			      || name[fnlen] == STRUCTS_SEPARATOR)) {
				next = (name[fnlen] != '\0') ?
				    name + fnlen + 1 : NULL;
				break;
			}
		}
		if (field->name == NULL) {
			errno = ENOENT;
			return (NULL);
		}
		type = field->type;
		data = (char *)data + field->offset;
		break;
	    }
	case STRUCTS_TYPE_UNION:
	    {
		const struct structs_ufield *const fields = type->args[0].v;
		const char *const mtype = type->args[1].s;
		struct structs_union *const un = (void *) data;
		const size_t oflen = strlen(un->field_name);
		const struct structs_ufield *ofield;
		const struct structs_ufield *field;
		void *new_un;
		void *data2;

		/* Special handling for "field_name" */
		if (strcmp(name, "field_name") == 0) {
			type = &structs_type_union_field_name;
			data = (void *)&un->field_name;
			break;
		}

		/* Find the old field */
		for (ofield = fields; ofield->name != NULL; ofield++) {
			if (strcmp(ofield->name, un->field_name) == 0)
				break;
		}
		if (ofield->name == NULL) {
			assert(0);
			errno = EINVAL;
			return (NULL);
		}

		/*
		 * Check if the union is already set to the desired field,
		 * handling field names with the separator char in them.
		 */
		if (strncmp(name, un->field_name, oflen) == 0
		    && (name[oflen] == '\0'
		      || name[oflen] == STRUCTS_SEPARATOR)) {
			next = (name[oflen] != '\0') ? name + oflen + 1 : NULL;
			field = ofield;
			goto union_done;
		}

		/* Is modifying the union to get the right name acceptable? */
		if (!set_union) {
			errno = ENOENT;
			return (NULL);
		}

		/* Find the new field */
		for (field = fields; field->name != NULL; field++) {
			const size_t fnlen = strlen(field->name);

			/* Handle field names with separator in them */
			if (strncmp(name, field->name, fnlen) == 0
			    && (name[fnlen] == '\0'
			      || name[fnlen] == STRUCTS_SEPARATOR)) {
				next = (name[fnlen] != '\0') ?
				    name + fnlen + 1 : NULL;
				break;
			}
		}
		if (field->name == NULL) {
			errno = ENOENT;
			return (NULL);
		}

		/* Create a new union with the new field type */
		if ((new_un = MALLOC(mtype, field->type->size)) == NULL)
			return (NULL);
		if ((*field->type->init)(field->type, new_un) == -1) {
			FREE(mtype, new_un);
			return (NULL);
		}

		/* See if name would be found with new union instead of old */
		data2 = new_un;
		if (next != NULL
		    && structs_find(field->type, next, (const void **) &data2, 1) == NULL) {
			(*field->type->uninit)(field->type, new_un);
			FREE(mtype, new_un);
			return (NULL);
		}

		/* Replace existing union with new one having desired type */
		(*ofield->type->uninit)(ofield->type, un->un);
		FREE(mtype, un->un);
		un->un = new_un;
		*((const char **)&un->field_name) = field->name;

	union_done:
		/* Continue recursing */
		type = field->type;
		data = un->un;
		break;
	    }
	default:
		assert(0);
		return (NULL);
	}

	/* Recurse on sub-element */
	if ((type = structs_find(type, next, (const void **) &data, set_union)) == NULL)
		return (NULL);

	/* Done */
	*datap = data;
	return (type);
}

/*
 * Traverse a structure.
 */

struct structs_trav {
	char		**list;
	u_int		len;
	u_int		alloc;
	const char	*mtype;
};

static int	structs_trav(struct structs_trav *t, const char *name,
			const struct structs_type *type, const void *data);

int
structs_traverse(const struct structs_type *type,
	const void *data, char ***listp, const char *mtype)
{
	struct structs_trav t;

	/* Initialize traversal structure */
	memset(&t, 0, sizeof(t));
	t.mtype = mtype;

	/* Recurse */
	if (structs_trav(&t, "", type, data) == -1) {
		while (t.len > 0)
			FREE(t.mtype, t.list[--t.len]);
		FREE(t.mtype, t.list);
		return (-1);
	}

	/* Return the result */
	*listp = t.list;
	return (t.len);
}

static int
structs_trav(struct structs_trav *t, const char *name,
	const struct structs_type *type, const void *data)
{
	const char *const dot = (*name == '\0' ? "" : ".");
	char *ename;
	int i;

	/* Dereference through pointer(s) */
	while (type->tclass == STRUCTS_TYPE_POINTER) {
		type = type->args[0].v;
		data = *((void **)data);
	}

	switch (type->tclass) {
	case STRUCTS_TYPE_PRIMITIVE:
	    {
		/* Grow list as necessary */
		if (t->len == t->alloc) {
			u_int new_alloc;
			char **new_list;

			new_alloc = (t->alloc + 32) * 2;
			if ((new_list = REALLOC(t->mtype,
			    t->list, new_alloc * sizeof(*t->list))) == NULL)
				return (-1);
			t->list = new_list;
			t->alloc = new_alloc;
		}

		/* Add new name to list */
		if ((t->list[t->len] = STRDUP(t->mtype, name)) == NULL)
			return (-1);
		t->len++;

		/* Done */
		return (0);
	    }

	case STRUCTS_TYPE_ARRAY:
	    {
		const struct structs_type *const etype = type->args[0].v;
		const struct structs_array *const ary = data;

		/* Iterate over array elements */
		for (i = 0; i < ary->length; i++) {
			const void *const edata
			    = (char *)ary->elems + (i * etype->size);

			if (ASPRINTF(t->mtype,
			    &ename, "%s%s%d", name _ dot _ i) == -1)
				return (-1);
			if (structs_trav(t, ename, etype, edata) == -1) {
				FREE(t->mtype, ename);
				return (-1);
			}
			FREE(t->mtype, ename);
		}

		/* Done */
		return (0);
	    }

	case STRUCTS_TYPE_FIXEDARRAY:
	    {
		const struct structs_type *const etype = type->args[0].v;
		const u_int length = type->args[2].i;

		/* Iterate over array elements */
		for (i = 0; i < length; i++) {
			const void *const edata
			    = (char *)data + (i * etype->size);

			if (ASPRINTF(t->mtype,
			    &ename, "%s%s%d", name _ dot _ i) == -1)
				return (-1);
			if (structs_trav(t, ename, etype, edata) == -1) {
				FREE(t->mtype, ename);
				return (-1);
			}
			FREE(t->mtype, ename);
		}

		/* Done */
		return (0);
	    }

	case STRUCTS_TYPE_STRUCTURE:
	    {
		const struct structs_field *field;

		/* Iterate over structure fields */
		for (field = type->args[0].v; field->name != NULL; field++) {
			const void *const edata
			    = (char *)data + field->offset;

			if (ASPRINTF(t->mtype,
			    &ename, "%s%s%s", name _ dot _ field->name) == -1)
				return (-1);
			if (structs_trav(t, ename, field->type, edata) == -1) {
				FREE(t->mtype, ename);
				return (-1);
			}
			FREE(t->mtype, ename);
		}

		/* Done */
		return (0);
	    }

	case STRUCTS_TYPE_UNION:
	    {
		const struct structs_ufield *const fields = type->args[0].v;
		const struct structs_union *const un = data;
		const struct structs_ufield *field;

		/* Find field */
		for (field = fields; field->name != NULL
		    && strcmp(un->field_name, field->name) != 0; field++);
		if (field->name == NULL) {
			assert(0);
			errno = EINVAL;
			return (-1);
		}

		/* Do selected union field */
		if (ASPRINTF(t->mtype,
		    &ename, "%s%s%s", name _ dot _ field->name) == -1)
			return (-1);
		if (structs_trav(t, ename, field->type, un->un) == -1) {
			FREE(t->mtype, ename);
			return (-1);
		}
		FREE(t->mtype, ename);

		/* Done */
		return (0);
	    }

	default:
		assert(0);
		errno = EDOM;
		return (-1);
	}
}

